### PR TITLE
fix(server): join on thread when Listening drops

### LINKED ISSUE #447: Main thread shouldn't panic when child threads panic
When running the following code, the panic in the child thread causes the server to crash:
```
extern crate hyper;
 
use hyper::Server;
use hyper::server::Request;
use hyper::server::Response;
use hyper::net::Fresh;
 
fn panic(_: Request, _: Response<Fresh>) {
    panic!("test")
}
 
fn main() {
    Server::http(panic).listen("127.0.0.1:3000").unwrap();
}
```
Here is the output:
```
thread '<unnamed>' panicked at 'test', src/main.rs:11
thread '<unnamed>' panicked at 'child thread None panicked', /home/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/thread/mod.rs:715
thread '<main>' panicked at 'child thread None panicked', /home/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/thread/mod.rs:715
An unknown error occurred

To learn more, run the command again with --verbose.
```
This is a fairly serious reliability issue, as a server should never go down because of its child threads.

### PR DESCRIPTION
**Pull request by seanmonstar**
Closes #447

### CODE REVIEW

File: src/server/mod.rs [hunk 1]
```diff
@@ -176,13 +176,24 @@ where S: NetworkStream + Clone, H: Handler {
...
+                let _ = guard.join();
+            },
+            None => ()
+        }
+    }
+}
+
 impl Listening {
     /// Stop the server from listening to its socket address.
     pub fn close(&mut self) -> HttpResult<()> {
```
Comment (reem) [inline src/server/mod.rs]: This function should probably take the guard for now, so that it at least lets you close a program running a server.

Comment (seanmonstar) [inline src/server/mod.rs]: ah right

**Comment by seanmonstar** [comment]
updated

File: src/server/mod.rs [hunk 2]
```diff
@@ -176,16 +176,28 @@ where S: NetworkStream + Clone, H: Handler {
...
     #[cfg(not(feature = "nightly"))]
-    _guard: JoinHandle,
+    _guard: Option<JoinHandle>,
     /// The socket addresses that the server is bound to.
     pub socket: SocketAddr,
 }
 
+impl Drop for Listening {
+    fn drop(&mut self) {
+        match self._guard.take() {
```
Comment (reem) [inline src/server/mod.rs]: much shorter to just write `self._guard.take().map(|g| let _ = g.join());`

**Comment by seanmonstar** [comment]
updated again

**Pull request merged**